File: gachisa-backend/src/main/java/com/gachisa/payment/repository/PaymentAttemptRepository.java
package com.gachisa.payment.repository;

import com.gachisa.payment.entity.PaymentAttempt;
import com.gachisa.payment.entity.PaymentAttemptStatus;
import java.time.LocalDateTime;
import java.util.Collection;
import java.util.List;
import java.util.Optional;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import jakarta.persistence.LockModeType;

public interface PaymentAttemptRepository extends JpaRepository<PaymentAttempt, Long> {

    Optional<PaymentAttempt> findByClientRequestId(String clientRequestId);

    Optional<PaymentAttempt> findByPgOrderId(String pgOrderId);

    @Query("select pa.paymentId from PaymentAttempt pa where pa.id = :attemptId")
    Optional<Long> findPaymentIdByAttemptId(@Param("attemptId") Long attemptId);

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("select pa from PaymentAttempt pa where pa.id = :attemptId")
    Optional<PaymentAttempt> findByIdForUpdate(@Param("attemptId") Long attemptId);

    Optional<PaymentAttempt> findFirstByPaymentIdOrderByCreatedAtDesc(Long paymentId);

    Optional<PaymentAttempt> findFirstByPaymentIdAndStatusOrderByCreatedAtDesc(
            Long paymentId, PaymentAttemptStatus status);

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    Optional<PaymentAttempt> findFirstByPaymentIdAndStatusInOrderByCreatedAtDesc(
            Long paymentId, Collection<PaymentAttemptStatus> statuses);

    List<PaymentAttempt> findTop100ByStatusAndNextRetryAtLessThanEqualOrderByNextRetryAtAsc(
            PaymentAttemptStatus status,
            LocalDateTime nextRetryAt
    );
}

File: gachisa-backend/src/main/kotlin/com/gachisa/payment/repository/PaymentAttemptRepository.kt
package com.gachisa.payment.repository

import com.gachisa.payment.entity.PaymentAttempt
import com.gachisa.payment.entity.PaymentAttemptStatus
import jakarta.persistence.LockModeType
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Lock
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import java.time.LocalDateTime
import java.util.Optional

interface PaymentAttemptRepository : JpaRepository<PaymentAttempt, Long> {
    fun findByClientRequestId(clientRequestId: String): Optional<PaymentAttempt>
    fun findByPgOrderId(pgOrderId: String): Optional<PaymentAttempt>

    @Query("select pa.paymentId from PaymentAttempt pa where pa.id = :attemptId")
    fun findPaymentIdByAttemptId(@Param("attemptId") attemptId: Long): Optional<Long>

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("select pa from PaymentAttempt pa where pa.id = :attemptId")
    fun findByIdForUpdate(@Param("attemptId") attemptId: Long): Optional<PaymentAttempt>

    fun findFirstByPaymentIdOrderByCreatedAtDesc(paymentId: Long): Optional<PaymentAttempt>

    fun findFirstByPaymentIdAndStatusOrderByCreatedAtDesc(
        paymentId: Long,
        status: PaymentAttemptStatus,
    ): Optional<PaymentAttempt>

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    fun findFirstByPaymentIdAndStatusInOrderByCreatedAtDesc(
        paymentId: Long,
        statuses: Collection<PaymentAttemptStatus>,
    ): Optional<PaymentAttempt>

    fun findTop100ByStatusAndNextRetryAtLessThanEqualOrderByNextRetryAtAsc(
        status: PaymentAttemptStatus,
        nextRetryAt: LocalDateTime,
    ): List<PaymentAttempt>
}
